### DISCUSSION TITLE
python/solved-how-to-get-python-requests-to-trust-a-self-signed-ssl-certificate/

### DISCUSSION THREAD

**Discussion by giscus** [CodeWithAbkhan/discussions-repo#4] [Announcements]
# python/solved-how-to-get-python-requests-to-trust-a-self-signed-ssl-certificate/

   
  Learn various methods to configure Python Requests to trust self-signed SSL certificates securely and effectively.
  
  
  

https://sqlpey.com/python/solved-how-to-get-python-requests-to-trust-a-self-signed-ssl-certificate/

**Comment by wahab-weeinvent** [comment]
Thanks for sharing the image links! Now, let's incorporate these into the guidance:

---

**How to install Flask through PowerShell:**

```powershell
pip install --isolated Flask
```

This command ensures installation without relying on environment variables and user configurations.

---

**For workplace-related issues, try this approach:**

```powershell
pip install -U <package_name> --user --proxy=<your proxy>
```

---

**To install from PyPI:**

```powershell
pip install -U -i https://pypi.org/simple package
```

**Explanation of flags:**

- `-U, --upgrade`: Upgrade all specified packages to the latest version.
- `-i, --index-url <url>`: Base URL of the Python Package Index (default: `pypi.org/simple`).

---

**If you encounter the ‘no matching version found’ error, follow this guide:**

1. Search your package online (e.g., search for `pycrypto pypi`).
2. Open the link on PyPI and download the files.
3. Open a Python shell and import an installed package to find its path:

```python
>>> import PIL
>>> PIL.__path__
['/Applications/MAMP/htdocs/canvas/src/zzd/env/lib/python3.7/site-packages/PIL']
```

The `PACKAGE.__path__` function shows where the packages are installed.

4. If you're unsure about installed packages, run `pip list` to see them.
5. Navigate to the path:

```sh
cd /Applications/MAMP/htdocs/canvas/src/zzd/env/lib/python3.7/site-packages/
```

6. Unzip the downloaded file and move it to the site-packages directory.
7. Navigate to the directory and run `setup.py` to install the package:

```sh
cd pycrypto-2.6.1
python setup.py install
```

You should now be able to use the package in Python.

---

**Obtaining the PEM file from a website:**

To download the PEM file from a trusted website, follow these steps:

1. Click on the lock icon next to the URL.

![Click on the lock](https://i.sstatic.net/sH50h.jpg)

2. Navigate to the certificates and open them.

![View certificates](https://i.sstatic.net/dh5Ch.jpg)

3. Download the PEM certificate chain.

![Download the PEM chain](https://i.sstatic.net/sH50h.jpg)

4. Place the `.PEM` file where your script can access it and verify with:

```python
r = requests.get(url, verify='path\to\pem_chain.pem')
```

---

**Additional Steps:**

1. **Using OpenSSL to retrieve certificates:**

```sh
echo | openssl s_client -servername NAME -connect HOST:PORT | sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' > certificate.crt
```

2. **Setting SSL certificate file path:**

```sh
export SSL_CERT_FILE=/path/file.crt
```

3. **Handling corporate network firewalls:**

```python
import os
os.environ["REQUESTS_CA_BUNDLE"] = 'path/to/corporate/cert.pem'
os.environ["SSL_CERT_FILE"] = 'path/to/corporate/cert.pem'
```

4. **Combining multiple root PEM files:**

```sh
$ cp myCert-A-Root.pem ca_roots.pem
$ cat myCert-B-Root.pem >> ca_roots.pem
$ echo "export REQUESTS_CA_BUNDLE=~/PATH_TO/CA_CHAIN/ca_roots.pem" >> ~/.bash_profile
```

---

I hope this helps! Let me know if you need any further assistance.
